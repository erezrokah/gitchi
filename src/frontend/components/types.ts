interface Author {
  avatarUrl: string;
  login: string;
}

interface User extends Author {
  name: string;
}

interface Comment {
  id: string;
  bodyText: string;
  createdAt: Date;
  author: Author;
  inReplyToId?: number;
  path?: string;
}

interface Channel {
  key: string;
  title: string;
  isReview: boolean;
  comments: Comment[];
}

interface Pr {
  id: string;
  bodyText: string;
  title: string;
  author: Author;
  channels: Channel[];
}
