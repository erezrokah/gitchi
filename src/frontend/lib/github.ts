export class AuthorizationError extends Error {}

const getToken = async () => {
  const tokenPromise: Promise<Record<string, string>> = new Promise(
    (resolve, reject) => {
      chrome.storage.sync.get(['token'], ({ token }) => {
        if (token) {
          resolve(token);
        } else {
          reject(new AuthorizationError());
        }
      });
    },
  );

  const token = await tokenPromise;

  return token;
};

const API_ENDPOINT = 'https://api.github.com/';

const get = async (path: string) => {
  const token = await getToken();
  const url = path.startsWith(API_ENDPOINT)
    ? path
    : `${API_ENDPOINT}${path}?timestamp=${Date.now()}`;
  const response = await fetch(url, {
    headers: {
      authorization: `token ${token}`,
    },
  });

  if (response.status === 401) {
    throw new AuthorizationError(response.statusText);
  } else if (!response.ok) {
    throw new Error(response.statusText);
  }

  return response;
};

const post = async (path: string, data = {}) => {
  const token = await getToken();
  const url = `${API_ENDPOINT}${path}`;
  const response = await fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      authorization: `token ${token}`,
    },
    body: JSON.stringify(data),
  });
  return response;
};

export const getCurrentUser = async () => {
  const user = await get('user');
  const { avatar_url: avatarUrl, login, name } = await user.json();
  return { avatarUrl, login, name };
};

export const parseLinkHeader = (link: string) => {
  const links = link.split(',').map(l => {
    const parts = l.split(';');

    const link = (parts[0] || '').trim().match(/<(.*?)>/);
    const rel = (parts[1] || '').match(/rel="(.*?)"/);

    return {
      link: link ? link[1] : '',
      rel: rel ? rel[1] : '',
    };
  });

  const result: Record<string, string> = {};

  links.forEach(({ rel, link }) => {
    result[rel] = link;
  });

  return result;
};

const getAllResponses = async (url: string) => {
  const maxResponses = 30;
  let responseCount = 1;

  const pageResponses = [];

  let nextURL = url;

  while (responseCount < maxResponses) {
    const pageResponse = await get(nextURL);
    pageResponses.push(pageResponse);
    responseCount++;

    const linkHeader = pageResponse.headers.get('Link');
    const next = linkHeader && parseLinkHeader(linkHeader)['next'];
    if (next) {
      nextURL = next;
    } else {
      break;
    }
  }

  const jsons = await Promise.all(pageResponses.map(r => r.json()));

  return jsons.flat();
};

interface IssueComment {
  id: number;
  body: string;
  created_at: Date;
  user: {
    login: string;
    avatar_url: string;
  };
}
interface PrComment extends IssueComment {
  in_reply_to_id?: number;
  path: string;
}

export const getReviewComments = async (
  owner: string,
  repo: string,
  prNumber: number,
) => {
  const result = await getAllResponses(
    `repos/${owner}/${repo}/pulls/${prNumber}/comments`,
  );

  const byReviewId: Record<string, Channel> = {};

  result.forEach(
    ({
      body: bodyText,
      id,
      created_at: createdAt,
      in_reply_to_id: inReplyToId,
      user,
      path,
    }: PrComment) => {
      const idStr = `${id}`;
      const comment = {
        bodyText,
        createdAt,
        id: idStr,
        inReplyToId,
        author: {
          avatarUrl: user.avatar_url,
          login: user.login,
        },
      } as Comment;

      if (inReplyToId && byReviewId[inReplyToId]) {
        byReviewId[inReplyToId].comments.push(comment);
      } else {
        byReviewId[id] = {
          key: idStr,
          title: path.replace(/^.*[\\\/]/, ''),
          isReview: true,
          comments: [comment],
        };
      }
    },
  );

  return Object.values(byReviewId);
};

export const getPullRequest = async (
  owner: string,
  repo: string,
  prNumber: number,
): Promise<Pr> => {
  const [pullRequest, commentsData, reviewComments] = await Promise.all([
    get(`repos/${owner}/${repo}/pulls/${prNumber}`).then(r => r.json()),
    get(`repos/${owner}/${repo}/issues/${prNumber}/comments`).then(r =>
      r.json(),
    ),
    getReviewComments(owner, repo, prNumber),
  ]);

  const comments = commentsData.map(
    ({ id, body: bodyText, user, created_at: createdAt }: IssueComment) => ({
      bodyText,
      createdAt,
      id: `${id}`,
      author: {
        avatarUrl: user.avatar_url,
        login: user.login,
      },
    }),
  );

  const pr = {
    id: `${pullRequest.id}`,
    bodyText: pullRequest.body,
    title: pullRequest.title,
    author: {
      avatarUrl: pullRequest.user.avatar_url,
      login: pullRequest.user.login,
    },
  };

  const result = {
    ...pr,
    channels: [
      {
        key: pr.id,
        title: 'Main',
        isReview: false,
        comments: [pr, ...comments],
      },
      ...reviewComments,
    ],
  };

  return result;
};

export const createPullRequestComment = async (
  owner: string,
  repo: string,
  prNumber: number,
  isReview: boolean,
  body: string,
  commentId: string,
) => {
  if (isReview) {
    const result = await post(
      `repos/${owner}/${repo}/pulls/${prNumber}/comments/${commentId}/replies`,
      { body },
    );
    console.log(await result.json());
  } else {
    const result = await post(
      `repos/${owner}/${repo}/issues/${prNumber}/comments`,
      { body },
    );
    console.log(await result.json());
  }
};
