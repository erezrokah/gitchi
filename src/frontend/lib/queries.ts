import gql from 'graphql-tag';

export const viewer = gql`
  query {
    viewer {
      login
      avatarUrl
      name
    }
  }
`;

export const pullRequest = gql`
  query pullRequest($repoOwner: String!, $repoName: String!, $prNumber: Int!) {
    repository(owner: $repoOwner, name: $repoName) {
      pullRequest(number: $prNumber) {
        id
        bodyText
        createdAt
        title
        author {
          avatarUrl
          login
        }
        comments(last: 50) {
          nodes {
            id
            bodyText
            createdAt
            author {
              avatarUrl
              login
            }
          }
        }
      }
    }
  }
`;
