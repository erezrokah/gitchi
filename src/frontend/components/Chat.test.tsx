/* eslint-disable @typescript-eslint/no-var-requires */
import * as React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import WS from 'jest-websocket-mock';
import {
  parseLocation,
  Chat,
  fetchCurrentUser,
  ActionType,
  reducer,
  onSendMessage,
  onDeleteMessage,
} from './Chat';
import { Pr, User } from './types';
import { AuthorizationError } from '../lib/github';

// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
// @ts-ignore
window.HTMLElement.prototype.scrollIntoView = jest.fn();

jest.mock('../lib/github', () => {
  const { AuthorizationError } = jest.requireActual('../lib/github');
  return {
    AuthorizationError,
    getCurrentUser: jest.fn(() => Promise.resolve()),
    getPullRequest: jest.fn(() => Promise.resolve()),
    createPullRequestComment: jest.fn(() => Promise.resolve()),
    deletePullRequestComment: jest.fn(() => Promise.resolve()),
  };
});

jest.mock('./Auth', () => {
  return {
    Auth: function auth({ onAuthSuccess }: { onAuthSuccess: jest.Mock }) {
      return (
        <div onClick={onAuthSuccess} data-testid="mocked-auth-component">
          Auth
        </div>
      );
    },
  };
});

jest.mock('../utils/webSocket', () => {
  return {
    getWebSocketUrl: jest.fn(() => Promise.resolve()),
  };
});

jest.spyOn(console, 'log').mockImplementation(() => {});

const location = { href: 'https://github.com/' };
// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
// @ts-ignore
global.window = Object.create(window);
Object.defineProperty(window, 'location', {
  value: location,
});

describe('Chat', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  describe('parseLocation', () => {
    test('should extract owner, repo and pr number from location href', () => {
      location.href = 'https://github.com/owner/repo/pull/666';

      expect(parseLocation()).toEqual({
        owner: 'owner',
        repo: 'repo',
        prNumber: 666,
      });
    });

    test('return null on none pull request url', () => {
      location.href = 'https://github.com/owner/repo/issues/111';

      expect(parseLocation()).toBeNull();
    });
  });

  const user: User = {
    name: 'name',
    avatarUrl: 'avatarUrl',
    login: 'login',
  };

  const author = {
    avatarUrl: 'authorAvatarUrl',
    login: 'authorLogin',
  };

  const pr: Pr = {
    id: '1',
    bodyText: 'bodyText',
    title: 'title',
    author,
    channels: [
      {
        key: '1',
        title: 'Main',
        isReview: false,
        comments: [
          {
            id: '100',
            bodyText: 'comment 100',
            createdAt: new Date(0),
            author,
            canDelete: false,
          },
        ],
      },
      {
        key: '2',
        title: 'Channel 2',
        isReview: true,
        comments: [
          {
            id: '200',
            bodyText: 'comment 200',
            createdAt: new Date(100),
            author: user,
            canDelete: true,
          },
        ],
      },
    ],
  };

  const {
    getCurrentUser,
    getPullRequest,
    createPullRequestComment,
    deletePullRequestComment,
  } = require('../lib/github') as {
    getCurrentUser: jest.Mock;
    getPullRequest: jest.Mock;
    createPullRequestComment: jest.Mock;
    deletePullRequestComment: jest.Mock;
  };

  const { getWebSocketUrl } = require('../utils/webSocket') as {
    getWebSocketUrl: jest.Mock;
  };

  describe('Chat', () => {
    afterEach(() => {
      WS.clean();
    });

    test('should match snapshot when loading', async () => {
      await act(async () => {
        const { asFragment } = render(<Chat />);
        expect(asFragment()).toMatchSnapshot();
        expect(getCurrentUser as jest.Mock).toHaveBeenCalledTimes(0);
      });
    });

    test('should match snapshot when not loading, but unauthenticated', async () => {
      await act(async () => {
        const userPromise = Promise.resolve();
        getCurrentUser.mockReturnValue(userPromise);
        const { asFragment, rerender } = render(<Chat />);

        rerender(<Chat />);

        await userPromise;

        expect(asFragment()).toMatchSnapshot();
        expect(getCurrentUser).toHaveBeenCalledTimes(1);
      });
    });

    test('should match snapshot when not loading, authenticated but no pr', async () => {
      await act(async () => {
        const userPromise = Promise.resolve(user);
        getCurrentUser.mockReturnValue(userPromise);

        const { asFragment, rerender } = render(<Chat />);

        rerender(<Chat />);

        await userPromise;

        expect(asFragment()).toMatchSnapshot();
        expect(getCurrentUser).toHaveBeenCalledTimes(1);
      });
    });

    test('should match snapshot when not loading, authenticated with pr', async () => {
      await act(async () => {
        location.href = 'https://github.com/owner/repo/pull/666';

        const userPromise = Promise.resolve(user);
        getCurrentUser.mockReturnValue(userPromise);

        const prPromise = Promise.resolve(pr);
        getPullRequest.mockReturnValue(prPromise);

        const { asFragment, rerender, getByText, queryByText } = render(
          <Chat />,
        );

        rerender(<Chat />);

        await userPromise;

        rerender(<Chat />);

        await prPromise;

        expect(asFragment()).toMatchSnapshot();
        expect(getCurrentUser).toHaveBeenCalledTimes(1);
        expect(getPullRequest).toHaveBeenCalledTimes(1);
        expect(getPullRequest).toHaveBeenCalledWith('owner', 'repo', 666);

        expect(
          getByText(pr.channels[0].comments[0].bodyText),
        ).toBeInTheDocument();
        expect(queryByText(pr.channels[1].comments[0].bodyText)).toBeNull();
      });
    });

    test('should match snapshot when collapsed', async () => {
      await act(async () => {
        location.href = 'https://github.com/owner/repo/pull/666';

        const userPromise = Promise.resolve(user);
        getCurrentUser.mockReturnValue(userPromise);

        const prPromise = Promise.resolve(pr);
        getPullRequest.mockReturnValue(prPromise);

        const { asFragment, rerender, getByTitle, queryByText } = render(
          <Chat />,
        );

        rerender(<Chat />);

        await userPromise;

        rerender(<Chat />);

        await prPromise;

        fireEvent.click(getByTitle('Collapse'));

        expect(asFragment()).toMatchSnapshot();

        expect(queryByText(pr.channels[0].comments[0].bodyText)).toBeNull();
        expect(queryByText(pr.channels[1].comments[0].bodyText)).toBeNull();
      });
    });

    test('should match snapshot channel is selected', async () => {
      await act(async () => {
        location.href = 'https://github.com/owner/repo/pull/666';

        const userPromise = Promise.resolve(user);
        getCurrentUser.mockReturnValue(userPromise);

        const prPromise = Promise.resolve(pr);
        getPullRequest.mockReturnValue(prPromise);

        const {
          asFragment,
          rerender,
          getByTitle,
          getByText,
          queryByText,
        } = render(<Chat />);

        rerender(<Chat />);

        await userPromise;

        rerender(<Chat />);

        await prPromise;

        fireEvent.click(getByTitle(pr.channels[1].title));

        expect(asFragment()).toMatchSnapshot();

        expect(
          getByText(pr.channels[1].comments[0].bodyText),
        ).toBeInTheDocument();
        expect(queryByText(pr.channels[0].comments[0].bodyText)).toBeNull();
      });
    });

    test('should call createPullRequestComment when send message button is clicked', async () => {
      await act(async () => {
        location.href = 'https://github.com/owner/repo/pull/666';

        const userPromise = Promise.resolve(user);
        getCurrentUser.mockReturnValue(userPromise);

        const prPromise = Promise.resolve(pr);
        getPullRequest.mockReturnValue(prPromise);

        const { getByTestId, rerender } = render(<Chat />);

        rerender(<Chat />);

        await userPromise;

        rerender(<Chat />);

        await prPromise;

        fireEvent.click(getByTestId('sendMessageButton'));

        expect(createPullRequestComment).toHaveBeenCalledTimes(1);
        expect(createPullRequestComment).toHaveBeenCalledWith(
          'owner',
          'repo',
          666,
          pr.channels[0].isReview,
          '',
          pr.channels[0].key,
        );
      });
    });

    test('should call deletePullRequestComment when delete button is clicked', async () => {
      await act(async () => {
        location.href = 'https://github.com/owner/repo/pull/666';

        const userPromise = Promise.resolve(user);
        getCurrentUser.mockReturnValue(userPromise);

        const prPromise = Promise.resolve(pr);
        getPullRequest.mockReturnValue(prPromise);

        const { getByTitle, rerender, getByTestId } = render(<Chat />);

        rerender(<Chat />);

        await userPromise;

        rerender(<Chat />);

        await prPromise;

        fireEvent.click(getByTitle(pr.channels[1].title));
        fireEvent.click(
          getByTestId(`delete-comment-${pr.channels[1].comments[0].id}`),
        );

        expect(deletePullRequestComment).toHaveBeenCalledTimes(1);
        expect(deletePullRequestComment).toHaveBeenCalledWith(
          'owner',
          'repo',

          pr.channels[1].isReview,
          pr.channels[1].comments[0].id,
        );
      });
    });

    test('should add listeners to web socket', async () => {
      await act(async () => {
        location.href = 'https://github.com/owner/repo/pull/666';

        const userPromise = Promise.resolve(user);
        getCurrentUser.mockReturnValue(userPromise);

        const prPromise = Promise.resolve(pr);
        getPullRequest.mockReturnValue(prPromise);

        const url = 'ws://localhost:1234';
        const webSocketUrlPromise = Promise.resolve(url);
        getWebSocketUrl.mockReturnValue(webSocketUrlPromise);
        const server = new WS('ws://localhost:1234');

        const { rerender } = render(<Chat />);

        rerender(<Chat />);

        await userPromise;

        rerender(<Chat />);

        await prPromise;

        rerender(<Chat />);

        await webSocketUrlPromise;

        expect(getWebSocketUrl).toHaveBeenCalledTimes(1);

        await server.connected;
        server.send('invalid json');
        server.close();

        expect(console.log).toHaveBeenCalledWith('GitHub WebSocket open');
        expect(console.log).toHaveBeenCalledWith(
          'error',
          new SyntaxError('Unexpected token i in JSON at position 0'),
        );
        expect(console.log).toHaveBeenCalledWith('GitHub WebSocket close');

        await expect(server).toReceiveMessage(
          `subscribe:pull_request:${pr.id}`,
        );
      });
    });

    test('should fetch pr data on pull_request update message', async () => {
      await act(async () => {
        location.href = 'https://github.com/owner/repo/pull/666';

        const userPromise = Promise.resolve(user);
        getCurrentUser.mockReturnValue(userPromise);

        const prPromise = Promise.resolve(pr);
        getPullRequest.mockReturnValue(prPromise);

        const url = 'ws://localhost:1234';
        const webSocketUrlPromise = Promise.resolve(url);
        getWebSocketUrl.mockReturnValue(webSocketUrlPromise);
        const server = new WS('ws://localhost:1234');

        const { rerender } = render(<Chat />);

        rerender(<Chat />);

        await userPromise;

        rerender(<Chat />);

        await prPromise;

        rerender(<Chat />);

        await webSocketUrlPromise;

        await server.connected;

        const data = [
          `pull_request:${pr.id}`,
          {
            timestamp: 5,
            wait: 200,
            reason: `pull request #${pr.id} updated`,
          },
        ];

        jest.clearAllMocks();
        jest.useFakeTimers();

        server.send(JSON.stringify(data));

        expect(console.log).toHaveBeenCalledWith(
          'GitHub Pull Request Updated',
          {
            id: pr.id,
            timestamp: 5,
            wait: 200,
            reason: `pull request #${pr.id} updated`,
          },
        );
        jest.advanceTimersByTime(300);
        expect(getPullRequest).toHaveBeenCalledTimes(1);
      });
    });

    test('should ignore non pull_request update messages', async () => {
      await act(async () => {
        location.href = 'https://github.com/owner/repo/pull/666';

        const userPromise = Promise.resolve(user);
        getCurrentUser.mockReturnValue(userPromise);

        const prPromise = Promise.resolve(pr);
        getPullRequest.mockReturnValue(prPromise);

        const url = 'ws://localhost:1234';
        const webSocketUrlPromise = Promise.resolve(url);
        getWebSocketUrl.mockReturnValue(webSocketUrlPromise);
        const server = new WS('ws://localhost:1234');

        const { rerender } = render(<Chat />);

        rerender(<Chat />);

        await userPromise;

        rerender(<Chat />);

        await prPromise;

        rerender(<Chat />);

        await webSocketUrlPromise;

        await server.connected;

        const data = [`some_other_message`];

        jest.clearAllMocks();
        jest.useFakeTimers();

        server.send(JSON.stringify(data));

        expect(console.log).not.toHaveBeenCalledWith(
          'GitHub Pull Request Updated',
        );
        jest.advanceTimersByTime(300);
        expect(getPullRequest).toHaveBeenCalledTimes(0);
      });
    });

    test('should fetch pr data when send refresh button is clicked', async () => {
      await act(async () => {
        location.href = 'https://github.com/owner/repo/pull/666';

        const userPromise = Promise.resolve(user);
        getCurrentUser.mockReturnValue(userPromise);

        const prPromise = Promise.resolve(pr);
        getPullRequest.mockReturnValue(prPromise);

        const { getByTitle, rerender } = render(<Chat />);

        rerender(<Chat />);

        await userPromise;

        rerender(<Chat />);

        await prPromise;

        jest.clearAllMocks();

        fireEvent.click(getByTitle('Refresh'));

        expect(getPullRequest).toHaveBeenCalledTimes(1);
        expect(getPullRequest).toHaveBeenCalledWith('owner', 'repo', 666);
      });
    });

    test('should call getCurrentUser when onAuthSuccess is invoked', async () => {
      await act(async () => {
        const userPromise = Promise.resolve();
        getCurrentUser.mockReturnValue(userPromise);
        const { getByTestId, rerender } = render(<Chat />);

        rerender(<Chat />);

        await userPromise;

        jest.clearAllMocks();

        fireEvent.click(getByTestId('mocked-auth-component'));

        expect(getCurrentUser).toHaveBeenCalledTimes(1);
      });
    });
  });

  describe('fetchCurrentUser', () => {
    test('should dispatch user received action on success', async () => {
      const user = {};
      const promise = Promise.resolve(user);
      getCurrentUser.mockReturnValue(promise);
      const dispatch = jest.fn();

      fetchCurrentUser(dispatch);
      await promise;

      expect(getCurrentUser).toHaveBeenCalledTimes(1);
      expect(dispatch).toHaveBeenCalledTimes(1);
      expect(dispatch).toHaveBeenCalledWith({
        type: ActionType.USER_RECEIVED,
        payload: user,
      });
    });

    test('should dispatch auth error action on auth error', async () => {
      const error = new AuthorizationError();
      const promise = Promise.reject(error);
      getCurrentUser.mockReturnValue(promise);
      const dispatch = jest.fn();

      fetchCurrentUser(dispatch);
      await promise.catch(() => {});

      expect(getCurrentUser).toHaveBeenCalledTimes(1);
      expect(dispatch).toHaveBeenCalledTimes(1);
      expect(dispatch).toHaveBeenCalledWith({
        type: ActionType.API_AUTH_ERROR_RECEIVED,
      });
    });

    test('should dispatch unknown error action on unknown error', async () => {
      const error = new Error();
      const promise = Promise.reject(error);
      getCurrentUser.mockReturnValue(promise);
      const dispatch = jest.fn();

      fetchCurrentUser(dispatch);
      await promise.catch(() => {});

      expect(getCurrentUser).toHaveBeenCalledTimes(1);
      expect(dispatch).toHaveBeenCalledTimes(1);
      expect(dispatch).toHaveBeenCalledWith({
        type: ActionType.API_UNKNOWN_ERROR_RECEIVED,
      });
    });
  });

  describe('reducer', () => {
    test(ActionType.USER_RECEIVED, () => {
      expect(
        reducer(
          { collapsed: false, loading: true },
          { type: ActionType.USER_RECEIVED, payload: user },
        ),
      ).toEqual({ collapsed: false, loading: false, user });
    });

    test(ActionType.API_AUTH_ERROR_RECEIVED, () => {
      expect(
        reducer(
          { collapsed: false, loading: true },
          { type: ActionType.API_AUTH_ERROR_RECEIVED },
        ),
      ).toEqual({ collapsed: false, loading: false });
    });

    test(ActionType.API_UNKNOWN_ERROR_RECEIVED, () => {
      expect(
        reducer(
          { collapsed: false, loading: true },
          { type: ActionType.API_UNKNOWN_ERROR_RECEIVED },
        ),
      ).toEqual({ collapsed: false, loading: false });
    });

    test(ActionType.MENU_TOGGLED, () => {
      expect(
        reducer(
          { collapsed: false, loading: true },
          { type: ActionType.MENU_TOGGLED, payload: true },
        ),
      ).toEqual({ loading: true, collapsed: true });
    });

    test(`${ActionType.PR_RECEIVED} initial load`, () => {
      expect(
        reducer(
          { collapsed: false, loading: false },
          { type: ActionType.PR_RECEIVED, payload: pr },
        ),
      ).toEqual({
        collapsed: false,
        loading: false,
        pr,
        selectedChannel: pr.channels[0].key,
      });
    });

    test(`${ActionType.PR_RECEIVED} selected channel removed`, () => {
      expect(
        reducer(
          { collapsed: false, loading: false, selectedChannel: 'Channel 1000' },
          { type: ActionType.PR_RECEIVED, payload: pr },
        ),
      ).toEqual({
        collapsed: false,
        loading: false,
        pr,
        selectedChannel: pr.channels[0].key,
      });
    });

    test(`${ActionType.PR_RECEIVED} selected channel not removed`, () => {
      expect(
        reducer(
          {
            collapsed: false,
            loading: false,
            selectedChannel: pr.channels[1].key,
          },
          { type: ActionType.PR_RECEIVED, payload: pr },
        ),
      ).toEqual({
        collapsed: false,
        loading: false,
        pr,
        selectedChannel: pr.channels[1].key,
      });
    });

    test(ActionType.CHANNEL_SELECTED, () => {
      expect(
        reducer(
          { collapsed: false, loading: false, selectedChannel: '1' },
          { type: ActionType.CHANNEL_SELECTED, payload: '2' },
        ),
      ).toEqual({ collapsed: false, loading: false, selectedChannel: '2' });
    });

    test('default', () => {
      const state = { collapsed: false, loading: false };
      expect(
        reducer(
          state,
          // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
          // @ts-ignore
          { type: 'Unknown' },
        ),
      ).toBe(state);
    });
  });

  describe('onSendMessage', () => {
    test('should not call createPullRequestComment when location parsing fails', async () => {
      location.href = 'https://github.com/owner/repo/issues/111';

      await onSendMessage({
        key: 'key',
        title: 'title',
        isReview: false,
        comments: [],
      })('message');

      expect(createPullRequestComment).toHaveBeenCalledTimes(0);
    });

    test('should not call createPullRequestComment when activeChannel is undefined', async () => {
      location.href = 'https://github.com/owner/repo/pull/666';

      await onSendMessage(undefined)('message');

      expect(createPullRequestComment).toHaveBeenCalledTimes(0);
    });
  });

  describe('onDeleteMessage', () => {
    test('should not call deletePullRequestComment when location parsing fails', async () => {
      location.href = 'https://github.com/owner/repo/issues/111';

      await onDeleteMessage({
        key: 'key',
        title: 'title',
        isReview: false,
        comments: [],
      })('message');

      expect(deletePullRequestComment).toHaveBeenCalledTimes(0);
    });

    test('should not call deletePullRequestComment when activeChannel is undefined', async () => {
      location.href = 'https://github.com/owner/repo/pull/666';

      await onDeleteMessage(undefined)('message');

      expect(deletePullRequestComment).toHaveBeenCalledTimes(0);
    });
  });
});
