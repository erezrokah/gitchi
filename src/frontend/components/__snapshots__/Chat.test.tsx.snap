// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat Chat should match snapshot channel is selected 1`] = `
<DocumentFragment>
  .c0 {
  width: 130px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  text-align: center;
  display: block !important;
}

.c1 {
  height: 500px;
  width: 500px;
  overflow: auto;
}

.c2 {
  min-width: 400px;
}

<div>
    <div>
      <div
        class="ui top attached tabular menu"
      >
        <a
          class="item c0"
          title="Main"
        >
          Main
        </a>
        <a
          class="active item c0"
          title="Channel 2"
        >
          Channel 2
        </a>
        <a
          class="item"
          title="Refresh"
        >
          <i
            aria-hidden="true"
            class="refresh icon"
          />
        </a>
        <a
          class="right item"
          title="Collapse"
        >
          <i
            aria-hidden="true"
            class="arrow down icon"
          />
          Collapse
        </a>
      </div>
    </div>
    <div
      class="ui bottom attached segment c1"
    >
      <div
        class="ui feed"
      >
        <div
          class="event"
        >
          <div
            class="label"
          >
            <img
              class="ui image"
              src="avatarUrl"
            />
          </div>
          <div
            class="content"
          >
            <div
              class="summary"
            >
              <a
                class="user"
              >
                login
              </a>
              <div
                class="text extra"
              >
                <div
                  class="markdown-body"
                >
                  <p>
                    comment 200
                  </p>
                </div>
              </div>
              <div
                class="date"
              >
                01/01/1970
              </div>
            </div>
            <div
              class="meta"
            >
              <a
                class="like"
              >
                <i
                  aria-hidden="true"
                  class="delete icon"
                  data-testid="delete-comment-200"
                />
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ui fluid action input c2"
    >
      <input
        name="message"
        placeholder="Message..."
        type="text"
        value=""
      />
      <button
        class="ui icon primary button"
        data-testid="sendMessageButton"
        title="Send Message"
      >
        <i
          aria-hidden="true"
          class="send icon"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Chat Chat should match snapshot when collapsed 1`] = `
<DocumentFragment>
  .c0 {
  width: 130px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  text-align: center;
  display: block !important;
}

<div>
    <div>
      <div
        class="ui top attached tabular menu"
      >
        <a
          class="active item c0"
          title="Main"
        >
          Main
        </a>
        <a
          class="item c0"
          title="Channel 2"
        >
          Channel 2
        </a>
        <a
          class="item"
          title="Refresh"
        >
          <i
            aria-hidden="true"
            class="refresh icon"
          />
        </a>
        <a
          class="right item"
          title="Open"
        >
          <i
            aria-hidden="true"
            class="arrow up icon"
          />
          Open
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Chat Chat should match snapshot when loading 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`Chat Chat should match snapshot when not loading, authenticated but no pr 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`Chat Chat should match snapshot when not loading, authenticated with pr 1`] = `
<DocumentFragment>
  .c0 {
  width: 130px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  text-align: center;
  display: block !important;
}

.c1 {
  height: 500px;
  width: 500px;
  overflow: auto;
}

.c2 {
  min-width: 400px;
}

<div>
    <div>
      <div
        class="ui top attached tabular menu"
      >
        <a
          class="active item c0"
          title="Main"
        >
          Main
        </a>
        <a
          class="item c0"
          title="Channel 2"
        >
          Channel 2
        </a>
        <a
          class="item"
          title="Refresh"
        >
          <i
            aria-hidden="true"
            class="refresh icon"
          />
        </a>
        <a
          class="right item"
          title="Collapse"
        >
          <i
            aria-hidden="true"
            class="arrow down icon"
          />
          Collapse
        </a>
      </div>
    </div>
    <div
      class="ui bottom attached segment c1"
    >
      <div
        class="ui feed"
      >
        <div
          class="event"
        >
          <div
            class="label"
          >
            <img
              class="ui image"
              src="authorAvatarUrl"
            />
          </div>
          <div
            class="content"
          >
            <div
              class="summary"
            >
              <a
                class="user"
              >
                authorLogin
              </a>
              <div
                class="text extra"
              >
                <div
                  class="markdown-body"
                >
                  <p>
                    comment 100
                  </p>
                </div>
              </div>
              <div
                class="date"
              >
                01/01/1970
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ui fluid action input c2"
    >
      <input
        name="message"
        placeholder="Message..."
        type="text"
        value=""
      />
      <button
        class="ui icon primary button"
        data-testid="sendMessageButton"
        title="Send Message"
      >
        <i
          aria-hidden="true"
          class="send icon"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Chat Chat should match snapshot when not loading, but unauthenticated 1`] = `
<DocumentFragment>
  <div
    data-testid="mocked-auth-component"
  >
    Auth
  </div>
</DocumentFragment>
`;
